Basic framework for writing and running CLI plugins

ijc opened the pull request:
**- What I did**

This is a small part of #1534, I wanted to get the basic framework out so that potential consumers could start to play with it and give feedback on how it works for them as well as getting some early feedback on the code.

NB this also includes the changes for #1558 since I was working on top of that.

There is still lots to do:
- [x] ~~I've left off the `docker cli-plugin *` stuff since it is still being discussed (although some bits of the internal scaffolding here for listing and finding plugins are more related to that than the actual functionality added here, but I didn't tease those apart).~~ Not doing this any more, at least not in this round (https://github.com/docker/cli/issues/1534#issuecomment-446921207).
- [x] List broken plugins as a new section in `docker help`.
- [ ] There are no docs;
- [ ] Although there are some here there are insufficient tests (both unit & e2e);
- [x] `docker someplugin --help` produces the top-level help instead of calling the plugin for some reason (this means one of the new e2e tests currently fails), looks like `docker someplugin --version` has the same problem, I'm investigating.
- [x] Expose plugins in `docker info`
- [x] Plugin search path configurable via `~/.docker/config.json`
- [ ] Apply the package naming the maintainers decide upon.
- [ ] All the stuff around exposing configuration files, credential handling, table formatting, using `dial-stdio` etc isn't here (Some of those are nops in practice and just need checking that they are usable externally, otherwise I plan to do these bits in one or more followups, somewhat based on/prioritise by the experiences of folks trying to use this). _edit_: to clarify, I don't intend on doing this bit here but rather in a followup.

**- How to verify it**

The unit and e2e tests cover some of it.

Plugin path can be configured with the "cliPluginsExtraDirs" key in `~/.docker/config.json` so for having a play (using a local config dir for isolation) you can also `make -f docker.Makefile build plugins` then things like:
```console
$ mkdir .docker-plugin-test
$ echo "{\"cliPluginsExtraDirs\": [\"`pwd`/build/plugins-linux-amd64/\"]}" > .docker-plugin-test/config.json
$ $ jq . .docker-plugin-test/config.json
{
  "cliPluginsExtraDirs": [
    ".../gocode/src/github.com/docker/cli/build/plugins-linux-amd64"
  ]
}
$ ./build/docker --config=.docker-plugin-test helloworld
```
to have a play.

**- Description for the changelog**

- Basic support for writing and running external CLI commands

vdemeester commented on cli-plugins/manager/candidate_test.go:
```diff
@@ -0,0 +1,81 @@
+	"gotest.tools/assert"
+)
+
+type mockCandidate struct {
```
absolute nit: this is more a `fake` than a `mock` :upside_down_face:

ijc replied:
I will run `sed`...

vdemeester commented on cli-plugins/manager/plugin.go:
```diff
@@ -0,0 +1,98 @@
+}
+
+// IsValid returns non-nil if the plugin is invalid/unusable.
+func (p *Plugin) IsValid() error {
```
nit: `IsValid` really wants to tell me it's a boolean (which is not in that case).

ijc replied:
I'm debating just dropping this and exposing `p.Err` instead. WDYT?

vdemeester commented on cmd/docker/docker.go:
```diff
@@ -32,14 +31,36 @@ func newDockerCommand(dockerCli *command.DockerCli) *cobra.Command {
 		SilenceErrors:    true,
 		TraverseChildren: true,
-		Args:             noArgs,
+		Args: func(_ *cobra.Command, _ []string) error {
```
Isn't this the same as no function provided ?

ijc replied:
Yes, I wasn't quite sure yet if I did actually need to do something here (relating to the issue with `--help` I mentioned in the intro). Will verify all that and drop if I can.

guillaumerose commented:
Just a comment related to the Docker Desktop integration: 
Docker Desktop doesn't change env variables, it's too hard (multiple shells possible, etc.). Would it be possible to also have a parameter representing `DOCKER_CLI_PLUGIN_EXTRA_DIRS` in `~/.docker/config.json` ?

ijc commented:
@vdemeester 

> why not github.com/docker/cli/plugins as a package name ?

To avoid confusion with `docker plugins` which is about daemon side plugins. Happy to take suggestions for better names though.

> could additionnal lookup path be available through the user configuration (aka $HOME/.docker/config.json) ?

Sure.

@guillaumerose WRT that setting and Desktop, we should try hard to avoid needing Desktop to set anything like this (whether in a file or the env) by making sure the default set of paths is sensible for the Desktop OSes such that desktop can install the base set of plugins in the right place without any special overrides.

vdemeester commented:
> To avoid confusion with `docker plugins` which is about daemon side plugins. Happy to take suggestions for better names though.

@ijc the `docker plugins` "code" is in `github.com/docker/cli/cli/command/plugin`, so I don't really think there would be any confusion (especially if we add a `README.md` in `github.com/docker/cli/plugins` to make it clear). cc @thaJeztah for thoughs

ijc commented on cli-plugins/examples/helloworld/main.go:
```diff
@@ -0,0 +1,25 @@
+		},
+		RunPlugin: func(cmd *cobra.Command, dockerCli command.Cli, args []string) {
+			fmt.Fprintln(dockerCli.Out(), "Hello World!")
+		},
```
This model for plumbing `dockerCli` to the command runs is completely broken for subcommands... I'm going to have to rethink. I'm going to try having `plugin.Run` take a `makeRoot func(command.Cli) cobra.Command` in place of the `plugin.Command`, I think that can work.

ijc commented:
I did a quick port of `docker app` to use the framework, mostly to serve as an example of how this might be used: https://github.com/docker/app/pull/433.

cquon commented on cli-plugins/plugin/plugin.go:
```diff
@@ -0,0 +1,96 @@
+		Short:            fullname + " is a Docker CLI plugin",
+		SilenceUsage:     true,
+		SilenceErrors:    true,
+		TraverseChildren: false,
```
When this is false it searches directly for the subcommand, so any plugin that has multiple layers of commands/subcommands, the parent flags will be ignored.
Example:

In my case normally, CLI command is
docker registry --host=1.1.1.1 <command>

which passes the host to all of the subcommands of the "registry" command.  Have verified that when this is false, it says the "host" flag is unrecognized, and it honors it when I changed this to "true".

ijc replied:
Thanks I think this is indeed wrong (or at least inconsistent wrt builtin type commands, which do use `true`). I'll flip this.

ijc commented on cli-plugins/plugin/plugin.go:
```diff
@@ -0,0 +1,96 @@
+		SilenceUsage:     true,
+		SilenceErrors:    true,
+		TraverseChildren: false,
+		PersistentPreRunE: func(cmd *cobra.Command, args []string) error {
```
I need to figure out the interaction between this and plugins which want to use `PersistentPreRun(E)` on their subcommand (or outlaw that, or document the issue).

ijc commented on e2e/cli-plugins/run_test.go:
```diff
@@ -0,0 +1,72 @@
+	res := icmd.RunCmd(icmd.Command("docker", "help", "helloworld"))
+	res.Assert(t, icmd.Expected{
+		ExitCode: 0,
+		Out:      "Usage:\n  docker helloworld",
```
I think this whole file ought really to be using golden files in preference to this simplistic mapping.

ijc commented:
I've force pushed quite a big update, rebasing on the latest code (so #1558 is no longer included since it is merged), sorted out the issues around `--help`, tidied and refactored a bunch of stuff, etc.

I'm going to update the todo in the first comment to reflect what's done and add some new entries people requested.

silvin-lubecki commented on Makefile:
```diff
@@ -42,10 +46,18 @@ cross: ## build executable for macOS and Windows
 	./scripts/build/osx
 
+.PHONY: plugins-osx
+plugins-osx: ## build example CLI plugins for for macOS
```
typo: double for

silvin-lubecki commented on cli-plugins/examples/badmeta/main.go:
```diff
@@ -0,0 +1,19 @@
+package main
```
I don't think this "example" adds much value, it should be an e2e test instead. My 2cts is that we should only have valid examples users can copy/paste to bootstrap their plugins.

ijc replied:
Actually this is only here to support the e2e testing, so you are right it's not really an example, I was just hoping to not need to battle the build system a second time to get it built.

So I'll move this one to e2e/cli-plugins/test-plugins and arrange (*somehow*) to have it built.

silvin-lubecki commented on cli-plugins/examples/helloworld/main.go:
```diff
@@ -0,0 +1,55 @@
+	"fmt"
+
+	cliplugins "github.com/docker/cli/cli-plugins"
+	"github.com/docker/cli/cli-plugins/plugin"
```
I didn't check the code and it's structure yet, but these imports (`docker/cli/cli-plugins` and especially `docker/cli/cli-plugins/plugin`) feel weird. Just having this first look, I think the Run function and the cliplugins.Metadata should be in the same package.

ijc replied:
IIRC that caused a circular include loop. I can certainly try and untangle it again though.

See also https://github.com/docker/cli/pull/1564#issuecomment-446963707 and the final bullet in the intro comment -- I'm awaiting MAINTAINERS guidance on the overall package naming.

ijc replied:
Moving metadata.go into `docker/cli/cli-plugins/plugin` in a quick and dirty way:
```
import cycle not allowed
package github.com/docker/cli/cmd/docker
	imports github.com/docker/cli/cli
	imports github.com/docker/cli/cli-plugins/manager
	imports github.com/docker/cli/cli-plugins/plugin
	imports github.com/docker/cli/cli
```

But moving it into `docker/cli/cli-plugins/manager` seems to have worked (at least the quick and dirty version), so I'll go with that for now (pending a renaming of the `cli-plugins` dir).

thaJeztah replied:
Is naming it `cli-plugin` potentially dangerous;

IIRC, Go treats

```go
import foobar.com/foo/package-name
```` 

the same as

```go
import name foobar.com/foo/package-name
```

(stripping the `package-` prefix, making it an implicit alias)

ijc replied:
IIRC you get whatever the `package foo` in the go files says which is distinct from the path you used to get to import it (customarily they are the same, of have a prefix stripped).

NB in the latest code there is nothing directly in `cli-plugins`, it's all subpackages so the thing you worry about doesn't crop up.

silvin-lubecki commented on cli-plugins/manager/plugin.go:
```diff
@@ -0,0 +1,145 @@
+// is set, and is always a `PluginError`, but the `Plugin` is still
+// returned with no error. An error is only returned due to a
+// non-recoverable error.
+func NewPlugin(c Candidate, rootcmd *cobra.Command) (Plugin, error) {
```
Does `NewPlugin` really needs to be exposed? It seems to be only used inside the `manager` package (or maybe I'm missing something)? And transitively I think Candidate shouldn't be exposed too.

ijc replied:
Yes, I think you are correct. Will fix.

silvin-lubecki commented on cli-plugins/manager/manager.go:
```diff
@@ -0,0 +1,190 @@
+)
+
+// ErrPluginNotFound is the error returned when a plugin could not be found.
+type ErrPluginNotFound string
```
I think we should expose a function `IsErrPluginNotFound` instead of exposing the error type itself. See https://dave.cheney.net/2014/12/24/inspecting-errors

ijc replied:
I think I can reuse `client.IsNotFound` perhaps?

simonferquel replied:
if your error type has a func NotFound(){} errdefs.ErrNotFound interface

silvin-lubecki commented on cli-plugins/manager/manager.go:
```diff
@@ -0,0 +1,190 @@
+func getPluginDirs(dockerCli command.Cli) []string {
+	pluginDirsOnce.Do(func() {
+		// Mostly for test.
+		if ds := os.Getenv("DOCKER_CLI_PLUGIN_EXTRA_DIRS"); ds != "" {
```
I'm not very comfortable with this code added only for tests. I'm concerned with security issues it might introduce, as it is loaded first, and might shadow regular plugins with corrupted ones? And as it is an undocumented environment variable and a non standard way to configure cli plugin directories, basic user will never notice it if it is set by someone else...

I think you need this env var only for e2e tests, but maybe you can transform you e2e test as a unit test, in this very package, from where then you can "patch" the `pluginsDirs` variable (even if it's not a good way too but somehow better I guess)?

ijc replied:
Despite the comment I'm actually considering whether this should perhaps be a user visible/usable/documented setting.

I don't think the e2e tests which need this can become unit tests, they need to actually run the CLI and see that it finds plugins etc.

ijc replied:
I thought I said this already, but I can't see where: I'm going to try and nuke the envvar by using `docker -C somecustomplace` in the e2e tests.

silvin-lubecki commented on cli-plugins/manager/manager_windows.go:
```diff
@@ -0,0 +1,11 @@
+)
+
+var defaultPluginDirs = []string{
+	filepath.Join(os.Getenv("USERPROFILE"), `.docker\cli-plugins`),
```
There's already a [homedir](https://github.com/moby/moby/tree/master/pkg/homedir) package in `moby/moby`, I don't know if it's worth using it. WDYT @thaJeztah ? I also spotted that there will be a new [os.UserHomeDir](https://tip.golang.org/doc/go1.12#os) function in `golang-1.12`. I don't know yet if the moby `homedir` package will be dropped or patched with the new standard function.

ijc replied:
The code which computes `~/.docker/config.json` uses the moby `homedir` function and I think this should therefore do the same -- will change.

ijc replied:
Speaking of `~/.docker/config.json`, looking at the code I think perhaps this ought to obey `$DOCKER_CONFIG` too, which probably motivates pushing this down to somewhere around `cli/config/config.go` and using the same code.

silvin-lubecki commented on cli-plugins/manager/manager.go:
```diff
@@ -0,0 +1,190 @@
+	pluginDirsOnce sync.Once
+)
+
+func getPluginDirs(dockerCli command.Cli) []string {
```
Just wondering, what about running this piece of code in an `init()` function instead of using this `sync.Once` ?

ijc replied:
Looking at this I'm not sure why this needs to be done exactly once, I don't think it's an especially hot path, it's not terribly expensive and I think the answer is idempotent in a given process (unless someone was futzing with its environment). Not sure what I was thinking. I think I should just nuke the sync stuff. WDYT?

silvin-lubecki replied:
I'm fine with the nuke 💥and compute it everytime.

ijc commented on cli-plugins/manager/cobra.go:
```diff
@@ -0,0 +1,58 @@
+// (optionally including invalid ones). The command stubs will have
+// several annotations added, see `CommandAnnotPlugin*`.
+func AddPluginCommandStubs(dockerCli command.Cli, cmd *cobra.Command, includeInvalid bool) error {
+	//fmt.Fprintf(os.Stderr, "Fall thru to HelpFunc\n")
```
I swear I fixed that once already locally. Will do it again!

silvin-lubecki replied:
🙄

cquon commented on cli/config/configfile/file.go:
```diff
@@ -48,6 +48,7 @@ type ConfigFile struct {
 	Experimental         string                      `json:"experimental,omitempty"`
 	StackOrchestrator    string                      `json:"stackOrchestrator,omitempty"`
 	Kubernetes           *KubernetesConfig           `json:"kubernetes,omitempty"`
+	CLIPluginExtraDirs   []string                    `json:"cli-plugin-extra-dirs,omitempty"`
```
Should we keep camel case convention for the config file?

ijc replied:
CLI is an abbreviation so I think this is camel case. Spelling it out as `CommandLineInterfacePluginExtraDirs` would make my hands hurt.

That said, I don't much like the name, plus I wonder if this ought to be a top level cli plugins object with a sub object with the paths array (and potentially in the future other things).

cquon replied:
Sorry I was talking about the json interpretation which is used for the configuration file parsing.
```
`json:"cli-plugin-extra-dirs,omitempty"`
```

but yes I do agree camel case would make it look strange, but I feel the same way about having everything follow a convention and then one with its own.

ijc replied:
oh, I misunderstood, hadn't noticed that the json bits were camelly. I guess it should be `cliPluginsExtraDirs` then (`cLIPLuginsExtraDirs` doesn't seem right...)

silvin-lubecki commented on cli-plugins/manager/manager.go:
```diff
@@ -0,0 +1,190 @@
+			return nil
+		}
+		// Portable? This seems to be as good as it gets :-/
+		if serr, ok := err.(*os.SyscallError); ok && serr.Err == syscall.ENOTDIR {
```
What about checking if the entry is a directory BEFORE the `ReadDir` ?
```go
 if info, err := os.Stat(d); err != nil || !info.IsDir() {
       return fmt.Errorf("failed to stat %q", d) 
 }
```

ijc replied:
Why didn't I think of that! (Perhaps trying to avoid ToCToU issues, but why!?!). I'll do as you suggest.

silvin-lubecki commented on cli-plugins/manager/manager.go:
```diff
@@ -0,0 +1,190 @@
+func listPluginCandidates(dirs []string) (map[string][]string, error) {
+	result := make(map[string][]string)
+	for _, d := range dirs {
+		if err := addPluginCandidatesFromDir(result, d); err != nil {
```
Maybe you can continue and append errors, so you can catch them as much as you can?

ijc replied:
Not sure about this, if there is a fatal error from `addPluginCandidatesFromDir` on one dir in the plugin search path then we aren't going to be doing much anyway so accumulating other subsequent errors doesn't seem all that useful and would mean using a slightly odd `[]error` return type or something (which would be ok if there was a compelling need to propagate multiple errors, but I'm not sure there is here).

silvin-lubecki commented on cli-plugins/manager/manager.go:
```diff
@@ -0,0 +1,190 @@
+}
+
+// ListPluginCandidates returns a map from plugin name to the list of (unvalidated) Candidates. The list is in descending order of priority.
+func ListPluginCandidates(dockerCli command.Cli) (map[string][]string, error) {
```
Does this function needs to be exposed? It feels more like it's an internal helper function?

silvin-lubecki commented on cli-plugins/manager/manager.go:
```diff
@@ -0,0 +1,190 @@
+		c := &candidate{paths[0]}
+		p, err := NewPlugin(c, rootcmd)
+		if err != nil {
+			return nil, err
```
Try to catch all errors here too?

ijc replied:
Only a fatal error would be returned here, if there is merely a problem with the candidate then that goes into `p.Err` but `err` is still `nil`.

silvin-lubecki commented on cli-plugins/manager/manager.go:
```diff
@@ -0,0 +1,190 @@
+}
+
+// FindPlugin finds a valid plugin, if the first candidate is invalid then returns an error
+func FindPlugin(dockerCli command.Cli, name string, rootcmd *cobra.Command, includeShadowed bool) (Plugin, error) {
```
Should this function be exported? It's only used in PluginRunCommand.

ijc replied:
I think it was used in `docker cli-plugins inspect`, but since I've now dropped that you are correct.

silvin-lubecki commented on cli-plugins/manager/manager.go:
```diff
@@ -0,0 +1,190 @@
+			plugin.ShadowedPaths = append(plugin.ShadowedPaths, path)
+		}
+	}
+	if plugin.Path == "" {
```
How can this case occur (except if getPluginDirs is nil...) ?

silvin-lubecki replied:
I mean, in the NewPlugin code, a plugin has always a path, otherwise it always returns an error, which feels sane.

ijc replied:
If `getPluginDirs` is non-nill but none of the directories in it contain the plugin then `NewPlugin` is not called and `plugin` is never initialized. Basically checking for `plugin.Path == ""` is a proxy for checking if `Plugin` was ever initialized, since as you say `NewPlugin` always sets that when it initialised the plugin.

silvin-lubecki commented on cli-plugins/manager/manager.go:
```diff
@@ -0,0 +1,190 @@
+	if err != nil {
+		return nil, err
+	}
+	if plugin.Err != nil {
```
This can be moved in `FindPlugin` no?

ijc replied:
Yes, now that `FindPlugin` only has a single caller I think the two can be collapsed.

ijc replied:
Furthermore, there are no longer any callers of `findPlugin` which pass `includeShadowed=true` which leads to a whole bunch of simplification and I think means `PluginRunCommand`→`runPluginCmmand`→`findPlugin` will sanely collapse into one function. I'll try that now.

silvin-lubecki commented on cli-plugins/manager/metadata.go:
```diff
@@ -0,0 +1,20 @@
+
+var (
+	// NamePrefix is the prefix required on all plugin binary names
+	NamePrefix = "docker-"
```
I think these can be constants instead.

ijc commented on cli-plugins/manager/manager_test.go:
```diff
@@ -0,0 +1,77 @@
+	assert.NilError(t, err)
+	exp := map[string][]string{
+		"plugin1": {
+			filepath.Join(dir.Path(), "plugins1", "docker-plugin1"),
```
Nice!

silvin-lubecki commented on cli-plugins/manager/metadata.go:
```diff
@@ -0,0 +1,20 @@
+	MetadataSubcommandName = "docker-cli-plugin-metadata"
+)
+
+// Metadata provided by the plugin
```
Maybe add more description here, as it is something plugin writers will have to use/fill (even if the fields are self-explanatory)?

ijc replied:
Would a ref to docs/extend/cli_plugins.md be enough?

silvin-lubecki replied:
I think it should be there for various reasons:
- godoc
- IDE auto-completion with documentation

silvin-lubecki commented on cli-plugins/manager/plugin.go:
```diff
@@ -0,0 +1,145 @@
+// field as a useful string and not just `{}`. See
+// https://github.com/golang/go/issues/10748 for some discussion
+// around why the builtin error type doesn't implement this.
+type PluginError struct {
```
I think this type shouldn't be exposed, nor `WrapAsPluginError` or `NewPluginError`.

ijc replied:
`NewPluginError` is used in cli/command/system/info_test.go, but I can unexport the other two I think.

silvin-lubecki commented on cli-plugins/manager/plugin.go:
```diff
@@ -0,0 +1,145 @@
+}
+
+// Plugin represents a potential plugin with all it's metadata.
+type Plugin struct {
```
nit: Move this type up in the file, as it is certainly the main type here users will read as a reference?

silvin-lubecki commented on cli-plugins/manager/plugin.go:
```diff
@@ -0,0 +1,145 @@
+		return Plugin{}, errors.Errorf("unable to determine basename of plugin candidate %q", path)
+	}
+	if runtime.GOOS == "windows" {
+		exe := ".exe"
```
nit: I guess this can be factorized with the 2 other places where there is a specific code path for windows?

ijc replied:
Two are the same but the other is different. I'll do what I can though.

silvin-lubecki commented on cli-plugins/manager/plugin.go:
```diff
@@ -0,0 +1,145 @@
+		}
+		fullname = strings.TrimSuffix(fullname, exe)
+	}
+	if !strings.HasPrefix(fullname, NamePrefix) {
```
Is this really needed? As candidates already have passed this test right?

silvin-lubecki commented on cli-plugins/manager/plugin.go:
```diff
@@ -0,0 +1,145 @@
+				continue
+			}
+			if cmd.Name() == p.Name || cmd.HasAlias(p.Name) {
+				p.Err = NewPluginError("plugin duplicates builtin command")
```
nit: add the `cmd` name in the error message, as it can be an alias and then be not so trivial to find?

ijc replied:
Good idea. I might even separate the non-alias from the alias case, then the message in one case could be `clashes with "foo" (an alias of "bar")`.

silvin-lubecki commented on cli-plugins/manager/plugin.go:
```diff
@@ -0,0 +1,145 @@
+		return p, nil
+	}
+
+	if p.Metadata.SchemaVersion != "0.1.0" {
```
You can use a library like `masterminds/semver`, check this to get inspiration https://github.com/docker/kamoulox-compose/blob/745afa4cf7ec811756c116c57a22470c561e6f8a/install/requirements.go#L33 😄

ijc replied:
As soon as there is a 0.2.0 etc that'll be useful, right now there is exactly one valid version so a string compare suffices. I've used github.com/blang/semver elsewhere too.

silvin-lubecki replied:
Yep `github.com/blang/semver` or `github.com/masterminds/semver` whatever, we just need to check if there isn't already one used somewhere in the CLI.
👍 to do that when a `0.2.0` will be there. But then maybe change the error message to something more clearer, like ` NewPluginError("plugin SchemaVersion %q is not valid, it must be 0.1.0", p.Metadata.SchemaVersion)`, it will be easier for plugin writers to catch that ?

silvin-lubecki commented on cli-plugins/plugin/plugin.go:
```diff
@@ -0,0 +1,128 @@
+	fullname := manager.NamePrefix + name
+
+	cmd := &cobra.Command{
+		Use:                   "docker" + " [OPTIONS] " + name + " [ARG...]",
```
nit: `fmt.Sprintf("docker [OPTIONS] %s [ARG...]", name)` works fine too 😄

silvin-lubecki commented on cli/cobra.go:
```diff
@@ -46,21 +94,68 @@ func FlagErrorFunc(cmd *cobra.Command, err error) error {
+				// have a single arg (the command) then try
+				// checking if it is a plugin and if it is
+				// then call it.
+				if len(args) == 1 && root.Annotations[annotPluginRootCommand] != "true" {
```
You can use the function declared just below:
```go
if len(args) == 1 && !isPlugin(root)
```

ijc replied:
Actually this went away entirely in the newest version.

silvin-lubecki commented on cli/command/cli.go:
```diff
@@ -166,6 +167,16 @@ func (cli *DockerCli) RegistryClient(allowInsecure bool) registryclient.Registry
 // Initialize the dockerCli runs initialization that must happen after command
 // line flags are parsed.
 func (cli *DockerCli) Initialize(opts *cliflags.ClientOptions) error {
+	cliflags.SetLogLevel(opts.Common.LogLevel)
```
I'm wondering why these changes are needed 🤔 ? Are they relevant for this PR? Or maybe do they fix a bug? In that case move it to another PR?

ijc replied:
See the first commit's commit message (7a9817d5b6476e1ee49c81f7861edc5605ed85fe):
```
    Fold `dockerPreRun` into `DockerCli.Initialize`
    
    All of the current callers follow the pattern:
    
        dockerPreRun(opts)
        err := dockerCli.Initialize(opts) ...
    
    So there is no semantic change into merging the content of `dockerPreRun` into the head of `Initialize`.
    
    I'm about to add a new caller outside of the `cmd/docker` package and this
    seems preferable exporting `DockerPreRun`.
    
    Signed-off-by: Ian Campbell <ijc@docker.com>
```

silvin-lubecki replied:
Right, seems legit 😄cc @thaJeztah / @vdemeester

ijc replied:
@thaJeztah @vdemeester any opinions?

thaJeztah replied:
Seems sensible; especially:

> this seems preferable exporting `DockerPreRun`.


(Even wondering if we need a separate `Initialize`, and if it should be part of the constructor)

ijc replied:
It needs be called separately from `PersistentPreRunE` because it needs to consume the parsed options but the `dockerCli` instance needs to exist before that so it can be captured into other scopes, such as `RunE`.

silvin-lubecki replied:
@vdemeester just spotted this https://godoc.org/github.com/spf13/cobra#OnInitialize
You should give it a try ?

ijc replied:
Some of the callers of this `Initialize` might be able to use `OnInitiliaze` (although it doesn't allow for errors to be returned) but I think (from code inspection only) there are other paths which would still need to manually call `Initialize`, at least `FlagErrorFunc` (I'm fairly confident) and possibly (I'm not sure) the `HelpFunc` stuff so I think it doesn't buy us a lot.

silvin-lubecki commented on cli/command/cli.go:
```diff
@@ -268,8 +279,8 @@ type ClientInfo struct {
 }
 
 // NewDockerCli returns a DockerCli instance with IO output and error streams set by in, out and err.
-func NewDockerCli(in io.ReadCloser, out, err io.Writer, isTrusted bool, containerizedFn func(string) (clitypes.ContainerizedClient, error)) *DockerCli {
```
Is this change really needed? I'm afraid it might break a lots of downstream projects... 😅

ijc replied:
It's worth mentioning that this is not the same interface as `github.com/docker/docker/client` which is the "programming API client" we'd expect those wanting to speak to the Docker API to be using and which includes flexibility wrt doing trust stuff.

This here is a (pretty much CLI internal) interface which is part of the docker/cli infrastructure/scaffolding, the consumers of this content trust setting are all in `docker/cli/cli/**/*.go` (callers of `dockerCli.ContentTrustEnabled()`). We do have some external users of this CLI scaffolding I think  -- but they are prime candidates for becoming plugins since their reason for using the docker/cli scaffolding is in essence to be compatible with the cli in some ways.

In any case it seems wrong to allow alternative CLI tools consuming the `docker/cli/cli/command` infrastructure to opt into or out of obeying the user's `$DOCKER_CONTENT_TRUST` setting.

Even if we do think it is a reasonable knob to give consumers I don't think making them manually decode `$DOCKER_CONTENT_TRUST` is the right way to approach it.

FWIW this was done in c15597947233a4a151dc5fa286e324c6ee0a6b4a:
```
    Push handling of `$DOCKER_CONTENT_TRUST` down into `NewDockerCli`
    
    I'm about to add a new (internal) user of this function and this saves
    exporting or duplicating then logic.
    
    Both the `docs/yaml` and `man` use don't do anything which would be affected by
    enabling/disabling content trust (nor would you expect them to) so switching
    from a hardcoded `false` to probing the envvar makes no practicaly difference.
    
    Note that this changes the API of `NewDockerCli`, but actual external callers
    should be unexpected.
    
    Signed-off-by: Ian Campbell <ijc@docker.com>
```

silvin-lubecki replied:
Ok I'm fine with the explanation, and you are right my concerns were about docker-app, which will be translated as a plugin. Any other concerns about that cc @thaJeztah @vdemeester ?

ijc replied:
I'm going to be handling this by rebasing onto #1633.

ijc commented on cmd/docker/docker.go:
```diff
@@ -92,9 +100,38 @@ func setFlagErrorFunc(dockerCli *command.DockerCli, cmd *cobra.Command, flags *p
+	if err != nil {
+		return err
+	}
+	helpcmd, err := pluginmanager.PluginRunCommand(dockerCli, cmd.Name(), root)
```
This is a subtle one. `PluginRunCommandHelp` is for precisely the case of running the plugin as `docker-foo help` (as a result of `docker help foo` or `docker foo help`).

This path is the `docker foo bar baz --help` path, ends up here (in the `HelpFunc` of the root command, so we need to call through to the plugin passing the actual `os.Args` (AKA `bar baz --help`) which is what `PluginRunCommand` does for us.

silvin-lubecki replied:
Hooooooooo right 👍

ijc replied:
In the end the distinction between `PluginRunCommand` and `PluginHelpCommand` went away and only the former remains.

silvin-lubecki commented on docs/extend/cli_plugins.md:
```diff
@@ -0,0 +1,83 @@
+
+### Naming
+
+A valid CLI plugin name consisists only of lower case letters `a-z`
```
consisists?

GordonTheTurtle commented:
<!-- AUTOMATED:POULE:DCO-EXPLANATION -->
Please sign your commits following these rules:
https://github.com/moby/moby/blob/master/CONTRIBUTING.md#sign-your-work
The easiest way to do this is to amend the last commit:
~~~console
$ git clone -b "plugins" git@github.com:ijc/docker-cli.git somewhere
$ cd somewhere
$ git rebase -i HEAD~842354607616
editor opens
change each 'pick' to 'edit'
save the file and quit
$ git commit --amend -s --no-edit
$ git rebase --continue # and repeat the amend for each commit
$ git push -f
~~~

Amending updates the existing PR. You **DO NOT** need to open a new one.

ijc commented:
Big force push addressing almost all of the review up to now, thanks @silvin-lubecki and @cquon.

Next step is to go through the comments and hit `resolve` to catch any stragglers.

ijc commented:
Is it possible for `dco-signed` to have flaked? I really don't see the bad commit...

ijc commented on cli-plugins/manager/manager_windows.go:
```diff
@@ -0,0 +1,11 @@
+
+var defaultPluginDirs = []string{
+	filepath.Join(os.Getenv("USERPROFILE"), `.docker\cli-plugins`),
+	`C:\ProgramData\Docker\cli-plugins`,
```
Not really sure what to do about this, given that Windows doesn't (appear to) have an equivalent "global" default akin to `/usr/foo` and `/usr/local/foo`. 

The engine has a bunch of what appear to be [hardcoded values](https://github.com/moby/moby/blob/master/container/container_windows.go#L15...L17) in the same path (if those are configurable my `git grep` is not finding it). 

It is possible to augment this set of paths via `~/.docker/config.json` though -- which I think is sufficient if things are installed in other non-default paths.

ijc replied:
I've honestly no idea, there were comments on the design which lead me to this particular set, perhaps @simonferquel you could advise?

Like with the other comment, there is a user specific one too, it's in `defaultUserPluginDir` here which is (now) `defaultSystemPluginDirs` but it isn't here because it is the same on both Unix and Windows (since it uses `github.com/docker/cli/cli/config` which already does whatever abstraction is needed)

simonferquel replied:
os.GetEnv("ProgramData")  should give you the shared program data dir (common to all users) with default value being c:\ProgramData (can change if you install Windows on a different drive than C)

cquon commented:
With latest git tag, it's not playing nicely with dep ensure dependency management (or maybe my config is wrong).  Looks like it's running into this issue:

https://github.com/golang/dep/issues/1306

Culprit is cli-plugins/ directory which only has nested directories no go files.  I tested with forking a branch with adding a dummy file (https://github.com/cquon/cli/tree/plugins/cli-plugins) and adding that to Gopkg.toml and I was able to update my package.

cquon commented on docs/extend/cli_plugins.md:
```diff
@@ -0,0 +1,83 @@
+---
```
Can we add information here on potentially default paths for detecting plugins?  Had to traverse the code to find it would be good to document probably.

ijc replied:
Yes, good idea.

cquon commented:
Will there be support to pass in flags into the subcommand plugins?
```
./docker-cli-plugin-new <plugin_cmd> ls <arg0> -q
```
throws an error unknown shorthand flag: 'q' in -q
directly it accepts the flag 

~/.docker/cli-plugins/docker-<plugin_cmd> <plugin_cmd> ls <arg0> -q

works

ijc commented:
@cquon 

> With latest git tag, it's not playing nicely with dep ensure dependency management

`dep` is a complete mystery to me, but I had something similar when I rebased https://github.com/docker/app/pull/433. I turned out to have a `"github.com/docker/cli/cli-plugins"` left in my imports somewhere, (because `Metadata` had moved). It went away after I fixed that.

> Also I believe "DOCKER_CLI_PLUGIN_EXTRA_DIRS" is no longer valid correct?

Correct. This has moved to `.docker/config.json` e.g. for hacking/testing I'm using:
```json
{
    "cliPluginsExtraDirs": [
	"/home/ijc/development/Docker/cli-plugins/gocode/src/github.com/docker/cli/build/plugins-linux-amd64",
	"/home/ijc/development/Docker/Build/gocode/src/github.com/docker/app/bin" ]
}
```

> Will there be support to pass in flags into the subcommand plugins?

Yes, that should work and I thought it did. I will have a look today.

ijc commented:
> > Will there be support to pass in flags into the subcommand plugins?

> Yes, that should work and I thought it did. I will have a look today.

I wrote a new e2e test and confirm this is a problem. On it.

thaJeztah commented on cli/command/cli.go:
```diff
@@ -419,8 +419,8 @@ type ClientInfo struct {
 // NewDockerCli returns a DockerCli instance with IO output and error streams set by in, out and err.
-func NewDockerCli(in io.ReadCloser, out, err io.Writer, isTrusted bool, containerizedFn func(string) (clitypes.ContainerizedClient, error)) *DockerCli {
-	return &DockerCli{in: NewInStream(in), out: NewOutStream(out), err: err, contentTrust: isTrusted, newContainerizeClient: containerizedFn}
+func NewDockerCli(in io.ReadCloser, out, err io.Writer, containerizedFn func(string) (clitypes.ContainerizedClient, error)) *DockerCli {
```
If we're changing this Exported function anyway, would it  make sense to rewrite it to use functional arguments?

Things like this start to get ugly;

```go
stdin, stdout, stderr := term.StdStreams()
dockerCli := command.NewDockerCli(stdin, stdout, stderr, nil)
```

And with functional arguments, it could look like, e.g. (we may want error-handling in there, but  just as a quick thought);

```go
func NewDockerCli(ops ...func(*DockerCli)) *DockerCli
```
```go
dockerCli := command.NewDockerCli(
	WithStandardStreams,
	WithContentTrust,
	WithContainerizedClient,
)
```

(this could make the interface more stable, as new options can be added without having to change the interface)

silvin-lubecki replied:
Just thought about this concrete example: `duffle` the CNAB client which uses the docker cli to pull images. See https://github.com/deislabs/duffle/blob/master/pkg/driver/docker_driver.go#L100
And I don't think `duffle` will ever be a docker cli plugin 😄

ijc replied:
AFAICS (and I only had a quick look) they only want a cli so they can call `cli.Client().ImagePull()` (and other similar constructs). They'd be as well to go at the `github.com/docker/docker/client` interfaces (which is effectively what `cli.Client()` gives you), the `cli` isn't really adding very much of use here.

And referring back to https://github.com/docker/cli/pull/1564/#discussion_r248236365 (which is what I tried to link to in my comment 2 ago but linked just to the PR by mistake), if we consider such use of the `NewDockerCli` to be fine, I still don't think it makes sense for other tools to just hardcode content trust to off.

thaJeztah commented on cli/cobra_test.go:
```diff
@@ -0,0 +1,30 @@
+	"gotest.tools/assert"
+)
+
+func TestVisitAll(t *testing.T) {
```
❤️ thanks!

ijc replied:
You're welcome!

ijc commented on cli-plugins/manager/manager.go:
```diff
@@ -0,0 +1,83 @@
+type notFound interface{ NotFound() }
+
+// IsNotFound is true if the given error is due to a plugin not being found.
+func IsNotFound(err error) bool {
```
My observation before I went this way was:
```
$ git grep func.IsNotFound
cli/manifest/store/store.go:func IsNotFound(err error) bool {
cli/registry/client/fetcher.go:func IsNotFound(err error) bool {
vendor/github.com/containerd/containerd/errdefs/errors.go:func IsNotFound(err error) bool {
vendor/github.com/docker/docker/errdefs/is.go:func IsNotFound(err error) bool {
vendor/k8s.io/apimachinery/pkg/api/errors/errors.go:func IsNotFound(err error) bool {
```
Consolidating them was a yakk I was hoping to avoid shaving in this already large PR ;-)

thaJeztah replied:
Definitely not a showstopper; perhaps we should have a look at moving that package to its own repo, and use it across the board.

ijc commented on cli-plugins/manager/manager_unix.go:
```diff
@@ -0,0 +1,8 @@
+package manager
+
+var defaultSystemPluginDirs = []string{
+	"/usr/local/lib/docker/cli-plugins", "/usr/local/libexec/docker/cli-plugins",
```
I wrote those on the same line because they are "cousins" (`lib` vs `libexec` for different Linux distros, the design has a footnote about that dichotomy).

This is the `defaultSystemPluginDirs`, there is also `defaultUserPluginDir` which is in the user's home dir (next to `config.json`, so obeying `--config` and in `.docker` by default etc). The user variant is not different on Windows so it isn't in this file.

ijc commented on cli-plugins/manager/plugin.go:
```diff
@@ -0,0 +1,96 @@
+				return p, nil
+			}
+			if cmd.HasAlias(p.Name) {
+				p.Err = errors.Errorf("plugin %q duplicates an alias of builtin command %q", p.Name, cmd.Name())
```
Yes, the whole "shadowing" notion deals with that, the first one we find wins and  the ones which come second onwards go into `ShadowedPaths` for visibility but are not otherwise usable/referencable.

ijc replied:
We raced, your edit was correct ;-)

ijc commented on cli-plugins/manager/plugin.go:
```diff
@@ -0,0 +1,96 @@
+		return Plugin{}, errors.Errorf("unable to determine basename of plugin candidate %q", path)
+	}
+	if runtime.GOOS == "windows" {
+		exe := ".exe"
```
For the case where we trim in `newPlugin` I suspect the answer is yes (will fix that).

Where we add it to execute I suspect the answer is not since the case doesn't matter then.

thaJeztah commented on cli-plugins/manager/manager.go:
```diff
@@ -41,6 +43,79 @@ func getPluginDirs(dockerCli command.Cli) []string {
+	result := make(map[string][]string)
+	for _, d := range dirs {
+		// Silently ignore any non-directories in the set of paths
+		if fi, err := os.Stat(d); err == nil && !fi.IsDir() {
```
Wondering if we should ignore errors due to permissions as well (iow; if a user does not have permissions on a directory, just skip that directory, and the plugins)

ijc replied:
That's coming close to skipping on any `Stat` error at all, maybe that's the right thing.

ijc replied:
I went with skipping anything which gave an error from `os.Stat` or wasn't a dir.

thaJeztah commented on cli-plugins/manager/candidate_test.go:
```diff
@@ -70,6 +72,7 @@ func TestValidateCandidate(t *testing.T) {
 			assert.ErrorContains(t, err, tc.err)
 		} else if tc.invalid != "" {
 			assert.NilError(t, err)
+			assert.Assert(t, cmp.ErrorType(p.Err, reflect.TypeOf(&pluginError{})))
```
Should we have an interface for plugin errors?

```go
type PluginError interface {
	Plugin()
}
```

ijc replied:
I'm not sure what that would buy us, only these unit tests really need to check and that's only because we want the `MarshalText` behaviour of `pluginError` (which normal `error` lacks, see the comment in `cli-plugins/manager/error.go`) for the `docker info --format '{{json .}}` case. No actual user/calling code should ever need to probe whether the `error` in `Plugin.Err` is one of these.

thaJeztah commented on cli/command/system/info_test.go:
```diff
@@ -21,8 +22,9 @@ func base64Decode(val string) []byte {
 	return decoded
 }
 
+var sampleID = "EKHL:QDUU:QZ7U:MKGD:VDXK:S27Q:GIPU:24B7:R7VT:DGN6:QCSF:2UBX"
```
nit: `const` ?

thaJeztah commented on cli/command/system/testdata/docker-info-plugins.golden:
```diff
@@ -0,0 +1,56 @@
+Insecure Registries:
+ 127.0.0.0/8
+Live Restore Enabled: false
+CLI Plugins:
```
Ok, thinking if we should (can) put this in the `docker info` output; `docker info` is not available if no connection to the daemon can be made.

If we _do_ want to integrate it, I think we should split it into `Server` and `Client` (similar to `docker version` showing two sections, and in case we cannot connect to the server, we should print the CLI section only.

ijc replied:
The suggestions to include in `docker info` came from @silvin-lubecki. With the errors being in `docker help` now we could perhaps get away without this, or else I could split as you suggest if you both agree with that approach.

silvin-lubecki replied:
I'm good with splitting the `docker info` command output into two sections. But I think the split should be done in another PR and rebase this one (which is already huge) on top of it. I spotted at least one info which is client only.
```
Debug Mode (client): false
```
WDYT?

ijc replied:
> in another PR 

That you are expecting me to do, right? (I'm ok with that, just want to know...)

> rebase this one (which is already huge)

There is a bunch of fairly generic preparatory cleanup which I've mostly floated to the front of the branch and could also easily (probably) pull out into another PR.

thaJeztah replied:
Alternatively, we can drop the "Integrate CLI plugins into `docker info`" commit from this PR, and move it to a new PR that makes those changes, but if that complicates things significantly, I'm ok with merging the current state, and move it in a follow up (Ideally. we should make sure to have it done before a release is cut though, otherwise it could potentially break people)

ijc replied:
Thinking about this:

> docker info is not available if no connection to the daemon can be made.

So I think in case the connection cannot be made the behaviour of the "pretty print" mode is clear enough -- you print an error instead of the server info.

But what about `-f {{json .}}` mode? Should it output a JSON object where all the server related things are null? Where should the error be reported (should it be?).

I presume the JSON output is an API and so cannot be changed, e.g. to push the server info down underneath a `server` key.

These seem like complicated questions to answer (and this little threadlet doesn't seem like the right context)

So would it be ok for me to just statically adjust the pretty version here to separate server and client side info but defer the changes to handle printing client side info when the server is not available to a followup?

ijc replied:
See https://github.com/docker/cli/pull/1638 for an attempt to deal with this stuff.

ijc replied:
Now merged, I'm rebasing onto that now (don't expect it before Monday though)

thaJeztah commented on cli/command/system/info.go:
```diff
@@ -207,9 +225,24 @@ func prettyPrintInfo(dockerCli command.Cli, info types.Info) error {
+				fmt.Fprintf(dockerCli.Out(), "  %s: (%s, %s) %s\n", p.Name, p.Version, p.Vendor, p.ShortDescription)
+			}
+		}
+		fmt.Fprintln(dockerCli.Out(), " invalid:")
```
Bit on the fence on the split between `valid` and `invalid`. Thinking if we can show them in the same list, given that we already print the error.

Actually; perhaps we should not list the invalid ones here at all, and add a warning for them to `info.Warnings`. WDYT?

thaJeztah replied:
Yes, I think putting it in the Warnings would be a great way to make them stand out without complicating this output. Invalid plugins would still be listed in the `--format {{.json}}`


Oh; just noticed this (and perhaps I should fix that for the Warnings as well); if there's no plugins, the value is `null`;

```json
{
  "Warnings": null,
  "CLIPlugins": null
}
```

Perhaps we should do a `Mirrors = make([]string, 0)` in case it's `nil`, so that it would become;

```json
{
  "Warnings": null,
  "CLIPlugins": []
}
```

ijc replied:
@thaJeztah Do you prefer `[]` to tagging it `omitempty` and skipping it altogether?

thaJeztah replied:
I should check, but I think having `[]` helps people that use the `--format` option; i.e. `--format '{{.CLIPlugins}}'` not producing an error if no plugin happens to be installed.

ijc replied:
Is there some trick to this?

In the `--format {{json .}}` output I'm getting `"Plugins":null` whether or not I include a 
```go
	if info.ClientInfo.Plugins == nil {
		info.ClientInfo.Plugins = make([]pluginmanager.Plugin, 0)
	}
```
(also tried `if len(...) == 0`). And `-f '{{.ClientInfo.Plugins}}'` works too no matter what.

ijc replied:
Nevermind, I was looking in the wrong place, adding the `make` *is* fixing the JSON.

I don't think `-f {{.ClientInfo.Plugins}}` was broken in the way you implied though, but it seems to work so that's fine by me ;-)

thaJeztah commented on cli-plugins/manager/metadata.go:
```diff
@@ -0,0 +1,19 @@
+)
+
+// Metadata provided by the plugin
+type Metadata struct {
```
As a follow-up would perhaps be nice to have a json-schema for this to define the spec. This could help plugin-authors to write a plugin in anything else than Go 🤔

ijc replied:
There is an _informal_ schema in `docs/extend/cli_plugins.md` (and right now there aren't many fields nor are they complex) but I do agree with you.

thaJeztah commented on docs/extend/cli_plugins.md:
```diff
@@ -0,0 +1,98 @@
+should not do so.
+
+Plugins distributed on Windows for system wide installation should be
+installed in `C:\ProgramData\Docker\cli-plugins`.
```
Just adding here so that we don't forget if we change the actual path to use `%PROGRAMDATA%`

ijc commented on cmd/docker/docker.go:
```diff
@@ -23,58 +22,79 @@ import (
+			// e.g. `docker plugin --foo` is caught here
+			// in the monolithic CLI and `foo` is reported
+			// as an unknown argument.
+			UnknownFlags: true,
```
https://github.com/spf13/pflag/pull/199 might allow an improvement to this. I guess it's not a blocker?

ijc commented:
I will rebase onto #1633 asap.

ijc commented:
@thaJeztah @silvin-lubecki I believe I've now addressed all of your comments (apart from the explicitly future work bits) and have just rebased onto latest master. It would be great to get any additional feedback (or an LGTM ;-)).

thaJeztah commented on cli/config/config.go:
```diff
@@ -46,6 +46,11 @@ func SetDir(dir string) {
 
+// Path returns the path to a file relative to the config dir
+func Path(p ...string) string {
+	return filepath.Join(append([]string{Dir()}, p...)...)
```
For a follow-up; thinking if we should consider preventing directory traversal in this utility (i.e., prevent the resulting path from being outside of the `config` directory)

thaJeztah commented on e2e/cli-plugins/testdata/docker-help-helloworld.golden:
```diff
@@ -1,8 +1,11 @@
 
-Usage:	docker helloworld COMMAND
+Usage:	docker helloworld [OPTIONS] COMMAND
```
nit: looks like `COMMAND` is optional for this plugin (the top-level `helloworld` is a functional command). 

```suggestion
Usage:	docker helloworld [OPTIONS] [COMMAND]
```


We could add a demo plugin that does not have a top-level command (so that it prints the `help` output if no subcommand is given)

(not a show-stopper, just something I noticed)

ijc replied:
> nit: looks like COMMAND is optional for this plugin (the top-level helloworld is a functional command).

Yeah, it looks like the `COMMAND` here comes from the existing `usageTemplate` so this would be presented the same way for a builtin command with similar behaviour.

ijc commented on cmd/docker/docker.go:
```diff
@@ -30,9 +31,28 @@ func newDockerCommand(dockerCli *command.DockerCli) *cobra.Command {
+			//      - https://github.com/golang/go/commit/d000e8742a173aa0659584aa01b7ba2834ba28ab
+			// os.Stdin is a *os.File which avoids this behaviour. We don't need the functionality
+			// of the wrappers here anyway.
+			plugincmd.Stdin = os.Stdin
```
I'm not sure if wrapping the `*os.File` into `io.ReadCloser` or `io.Writer` as `term.StdStreams()` would be susceptible to the issue mentioned in the comment, but in any case I think it's not really necessary here, the monolithic cli will have called `term.StdStreams()` before now anyway (so the things it does have actually been applied to these fds, albiet indirectly) but more importantly the plugin will (with the framework here) call `term.StdStreams` itself for its own output. I think it is correct to leave that decision up to the plugin.

thaJeztah replied:
makes sense; thx!

ijc commented on cli/cobra.go:
```diff
@@ -138,6 +141,21 @@ func wrappedFlagUsages(cmd *cobra.Command) string {
+	width := 13
+	if v, ok := cmd.Annotations[pluginmanager.CommandAnnotationPluginVendor]; ok {
+		if len(v) > width-2 {
+			v = v[:width-3] + "…"
```
Yes the `-3` is correct, it's the ellipsis (one char as you say) plus the surrounding `()`.

Your reference looks useful though, and tellingly it is more complicated (AKA complete) which somewhat suggests I've missed some corner cases...

thaJeztah replied:
Ah, did not take the surrounding `()` into account 😊 

> Your reference looks useful though, and tellingly it is more complicated (AKA complete) which somewhat suggests I've missed some corner cases...

Don't think it's a blocker, but we could switch to using that implementation (in a follow-up if you want)

ijc commented on cmd/docker/docker.go:
```diff
@@ -111,6 +111,21 @@ func setupHelpCommand(dockerCli *command.DockerCli, rootCmd, helpCmd *cobra.Comm
+		if len(args) > 0 {
+			helpcmd, err := pluginmanager.PluginRunCommand(dockerCli, args[0], rootCmd)
+			if err == nil {
+				helpcmd.Stdin = os.Stdin
```
Ack, will dup the comment.

thaJeztah replied:
Thx!

ijc replied:
Actually,there are three places doing this, but the right answer is to push the Stdio setup (and the comment) down into `PluginRunCommand` so I shall do that

ijc commented on cli-plugins/examples/helloworld/main.go:
```diff
@@ -18,16 +19,33 @@ func main() {
+			Short: "Print the API version of the server",
+			RunE: func(_ *cobra.Command, _ []string) error {
+				cli := dockerCli.Client()
+				ping, err := cli.Ping(context.Background())
```
The only reason this subcommand is here is to test the `PersistentPreRunE` arrangements (which is what causes `dockerCli.Client()` to be initialised) as part of 75145cb8b30b751e49f6ca370a61c15199b97296.

The actual interaction with the client version api is completely incidental so I don't think it is worth going any further with it.

thaJeztah replied:
:+1:

ijc commented on cli/command/system/testdata/docker-info-plugins.golden:
```diff
@@ -0,0 +1,56 @@
+Client:
+ Debug Mode: false
+ Plugins:
+  goodplugin: (0.1.0, ACME Corp) unit test is good
```
vendor and version or just vendor? I guess you mean the former so:
```
goodplugin: unit test is good (0.1.0, ACME Corp)
```
?

Alternative maybe:
```
goodplugin (0.1.0): unit test is good (ACME Corp)
```

ijc replied:
@thaJeztah with #1675 being merged I think for consistency:
```
goodplugin: unit test is good (ACME Corp, 0.1.0)
```
is the way to go. I'll make a PR.

ijc commented on cli/cobra.go:
```diff
@@ -178,7 +196,7 @@ Options:
 
 {{- range managementSubCommands . }}
-  {{rpad .Name .NamePadding }} {{.Short}}
+  {{rpad .Name .NamePadding }} {{ if isFirstLevelCommand .}}{{commandVendor .}} {{ end}}{{.Short}}
```
With the current layout the presence of a `(vendor)` between the name and the description (and the consequent alignment) makes the plugins stand out quite clearly and it's fairly obvious what is going on:

```
Commands:
  assemble                       (Docker Inc.) assemble is a high-level build tool
  attach                                       Attach local standard input, output, and error streams to a running container
  build                                        Build an image from a Dockerfile
  commit                                       Create a new image from a container's changes
[...]
  export                                       Export a container's filesystem as a tar archive
  helloworld                     (Docker Inc.) A basic Hello World plugin for tests
  history                                      Show the history of an image
```

Moving the vendor to the end would lose that alignment which makes things so clear, so you need/want the `*` but then it's not quite so obvious what that `*` actually indicates without other context.

thaJeztah replied:
>  makes the plugins stand out quite clearly 

That might be actually what I was looking at; (I think) the ultimate goal is to make plugins "feel" as if they were part of the actual CLI. So for most situations, the user wouldn't be interested if a command is a native command, or provided by a plugin.

A non-obtrusive indication may be sufficient for the day-to-day use, and all information would still be there for situations where the user would _actually_ have to deal with the plugin itself (e.g., verify if the correct version is installed)


Changing/tweaking this output should not be a breaking change, and I don't want to delay this PR any further though, so I'll create a list of things to discuss post-merge (the general functionality is more important)

ijc commented on docs/extend/cli_plugins.md:
```diff
@@ -0,0 +1,98 @@
+
+A plugin is required to support all of the global options of the
+top-level CLI, i.e. those listed by `man docker 1` with the exception
+of `-v`.
```
I fear cobra may be doing something tricky and short-circuiting through to the help command. I'll have a look.

thaJeztah replied:
Quick test; looks like this is because we add the parent command (which is the actual command that's being called) this seems to work, but we'd have to check if that's the right thing to do;

```patch
diff --git a/cli-plugins/examples/helloworld/main.go b/cli-plugins/examples/helloworld/main.go
index cbe01593..95dd2f93 100644
--- a/cli-plugins/examples/helloworld/main.go
+++ b/cli-plugins/examples/helloworld/main.go
@@ -44,6 +44,7 @@ func main() {
                        Run: func(cmd *cobra.Command, args []string) {
                                fmt.Fprintf(dockerCli.Out(), "Hello %s!\n", who)
                        },
+                       Version: "0.0.2",
                }
                flags := cmd.Flags()
                flags.StringVar(&who, "who", "World", "Who are we addressing?")
diff --git a/cli-plugins/plugin/plugin.go b/cli-plugins/plugin/plugin.go
index c72a68bb..5745b2aa 100644
--- a/cli-plugins/plugin/plugin.go
+++ b/cli-plugins/plugin/plugin.go
@@ -91,6 +91,10 @@ func newPluginCommand(dockerCli *command.DockerCli, plugin *cobra.Command, meta
        }
        opts, flags := cli.SetupPluginRootCommand(cmd)
 
+       if plugin.Version != "" {
+               cmd.Version = plugin.Version
+       }
+       cmd.SetVersionTemplate(plugin.Name() + " version {{.Version}}\n")
        cmd.SetOutput(dockerCli.Out())
 
        cmd.AddCommand(
```

ijc replied:
Bigger issue (that I'm invesitgating now) is that `docker helloworld -v` gets short circuited by the cli so the plugin never gets an opportunity to run. Once I've sorted that I'll figure out what needs to be done in the plugin framework (you suggestions looks like a good start)

ijc commented on docs/extend/cli_plugins.md:
```diff
@@ -0,0 +1,98 @@
+Plugins distributed on Windows for system wide installation should be
+installed in `%PROGRAMDATA%\Docker\cli-plugins`.
+
+User's may on all systems install plugins into `~/.docker/cli-plugins`.
```
As you (and the other maintainers) prefer.

thaJeztah replied:
I slightly prefer `~/.docker/plugins`, but I don't think it's a huge deal if we keep this for a follow-up as well (but decided on before GA 😅)

thaJeztah commented:
Summarizing things that I think can be handled in follow-ups (but some of these should be done before GA);

- [ ] https://github.com/docker/cli/pull/1564#discussion_r251818285 handling of `docker someplugin --version`. The `--version` / `-v` flag is handled by the plugin root-command, and therefore shows the `docker` version information (`Docker version 19.03.0-dev, build c8f1f1ee`). We should fix this to be handled by the `plugin`.
- [ ] Related to the above; consider making `--version` / `-v` part of the pluging "API", so that a plugin can advertise its version (both when calling the plugin directly (`docker-myplugin --version`), and when calling through the `docker` cli (`docker myplugin --version`). Perhaps we should add a standard function and/or define a standard format (`docker-myplugin version X, commit deadbeef (ACME, Inc.)`)
- [ ] https://github.com/docker/cli/pull/1564#discussion_r251819147 install location for plugins; choose between `~/.docker/plugins` or `~/.docker/cli-plugins`
- [ ] https://github.com/docker/cli/pull/1564#discussion_r251816130 decide on the output format in `docker --help`: make plugin-command stand out from built-in commands, or make them appear as regular commands with a non-obtrusive indicator (`*`)
- [ ] https://github.com/docker/cli/pull/1564#discussion_r251814998 decide on the location to show version and vendor for plugins in in `docker info`
- [ ] https://github.com/docker/cli/pull/1564#discussion_r251890532 small nit: use `displayutils.Ellipsis()` instead of a custom implementation.
- [ ] https://github.com/docker/cli/pull/1564#discussion_r251822916 small nit: hardening `config.Path()` to disallow directory traversal
- [ ] https://github.com/docker/cli/pull/1564#discussion_r251825582 small nit: fix usage output of the sample plugin

Future follow-ups (not discussed here);

- Extend the plugin framework to allow generating YAML-docs (https://github.com/docker/cli/tree/master/docs/yaml)
- Extend the plugin framework to allow generating man-pages (https://github.com/docker/cli/tree/master/man)
- Provide guidelines for bash-completion scripts (perhaps _generated_ for now), and work out how we can integrate those with the main bash-completion scripts, while remaining separate.

thaJeztah commented:
@docker/core-cli-maintainers looking at the list above, I'm inclined to move this forward, and take care of the listed issues separate; WDYT?

silvin-lubecki commented:
@thaJeztah I'm ok with merging this first plugin base and iterate with follow-ups on the issues you listed 👍

vdemeester commented:
Same as @silvin-lubecki given we do those follow-up before next release :angel:

ijc commented:
This WFM (of course) and I will commit to addressing the bullets in https://github.com/docker/cli/pull/1564/#issuecomment-458964609 next (my first act would be to transliterate that comment into an issue I think).

@thaJeztah I've looked over the outstanding comments and I think your list covers them all. Have I missed anything which you don't want to defer?

albers commented:
> do you have any advice on how we might go about integrating completion for plugins?

@ijc I think this is possible in the way you outlined.
I'll give it a try.

ijc commented:
Awesome, thanks @albers! Please give a shout if there is any way I can help
